File: packages/db-schema/src/schema/calendar-bindings.ts
import { sql } from 'drizzle-orm'
import { sqliteTable, text, index, uniqueIndex } from 'drizzle-orm/sqlite-core'
import type { VectorClock } from '@memry/contracts/sync-api'

export type CalendarBindingSourceType = 'event' | 'task' | 'reminder' | 'inbox_snooze'
export type CalendarBindingOwnershipMode = 'memry_managed' | 'provider_managed'
export type CalendarBindingWritebackMode = 'schedule_only' | 'time_and_text' | 'broad'

export const calendarBindings = sqliteTable(
  'calendar_bindings',
  {
    id: text('id').primaryKey(),
    sourceType: text('source_type').$type<CalendarBindingSourceType>().notNull(),
    sourceId: text('source_id').notNull(),
    provider: text('provider').notNull(),
    remoteCalendarId: text('remote_calendar_id').notNull(),
    remoteEventId: text('remote_event_id').notNull(),
    ownershipMode: text('ownership_mode').$type<CalendarBindingOwnershipMode>().notNull(),
    writebackMode: text('writeback_mode').$type<CalendarBindingWritebackMode>().notNull(),
    remoteVersion: text('remote_version'),
    lastLocalSnapshot: text('last_local_snapshot', { mode: 'json' }).$type<
      Record<string, unknown> | null
    >(),
    archivedAt: text('archived_at'),
    clock: text('clock', { mode: 'json' }).$type<VectorClock>(),
    syncedAt: text('synced_at'),
    createdAt: text('created_at')
      .notNull()
      .default(sql`(strftime('%Y-%m-%dT%H:%M:%fZ', 'now'))`),
    modifiedAt: text('modified_at')
      .notNull()
      .default(sql`(strftime('%Y-%m-%dT%H:%M:%fZ', 'now'))`)
  },
  (table) => [
    uniqueIndex('idx_calendar_bindings_source').on(table.sourceType, table.sourceId, table.provider),
    uniqueIndex('idx_calendar_bindings_remote').on(
      table.provider,
      table.remoteCalendarId,
      table.remoteEventId
    ),
    index('idx_calendar_bindings_source_type').on(table.sourceType),
    index('idx_calendar_bindings_archived_at').on(table.archivedAt)
  ]
)

export type CalendarBinding = typeof calendarBindings.$inferSelect
export type NewCalendarBinding = typeof calendarBindings.$inferInsert

File: apps/desktop-tauri/src-tauri/src/db/calendar_bindings.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
#[serde(rename_all = "camelCase")]
pub struct CalendarBinding {
    pub id: String,
    pub source_type: String,
    pub source_id: String,
    pub provider: String,
    pub remote_calendar_id: String,
    pub remote_event_id: String,
    pub ownership_mode: String,
    pub writeback_mode: String,
    pub remote_version: Option<String>,
    pub last_local_snapshot: Option<String>,
    pub archived_at: Option<String>,
    pub clock: Option<String>,
    pub synced_at: Option<String>,
    pub created_at: String,
    pub modified_at: String,
}

impl CalendarBinding {
    pub fn from_row(row: &rusqlite::Row<'_>) -> rusqlite::Result<Self> {
        Ok(Self {
            id: row.get("id")?,
            source_type: row.get("source_type")?,
            source_id: row.get("source_id")?,
            provider: row.get("provider")?,
            remote_calendar_id: row.get("remote_calendar_id")?,
            remote_event_id: row.get("remote_event_id")?,
            ownership_mode: row.get("ownership_mode")?,
            writeback_mode: row.get("writeback_mode")?,
            remote_version: row.get("remote_version")?,
            last_local_snapshot: row.get("last_local_snapshot")?,
            archived_at: row.get("archived_at")?,
            clock: row.get("clock")?,
            synced_at: row.get("synced_at")?,
            created_at: row.get("created_at")?,
            modified_at: row.get("modified_at")?,
        })
    }
}
